Mathlib/Topology/Algebra/Nonarchimedean/BasesNew.lean

```diff
@@ -0,0 +1,188 @@
+have some extra algebraic structure. For example, `Filter.IsAddGroupBasis.mk_of_subgroups_of_comm`
+shows that, in a commutative additive group, any filter basis made of subgroups is automatically
+a group filter basis.
+
+We also prove some extra results about the topological objects obtained in this setting
+(that is, nonarchimedean objects), but these should probably be tweaked to *not* mention
+`IsGroupBasis` (and friends) and moved elsewhere. More details in comments around the file.
+
+## Implementation note
+
+* We use subobject classes, so for example `Filter.IsAddGroupBasis.mk_of_subgroups` will
+  also apply if your basis is valued in `Submodule`, `Ideal`, ...
+
+* For a long time, each additional constructor in this file was its own predicate (e.g
+  there was a definition for "ring filter bases made of subgroups", another one for
+  "ring filter bases made of submodules", and so on...). Furthermore, these did not take any
+  predicate bounding the filter basis, making them inconsistent with both `Filter.IsBasis`
+  and `FilterBasis` (recall that the general setup used to rely on `FilterBasis`).
+  This led to a lot of (inconsistent) code duplication, which doesn't exist anymore.
+
+-/
+
+open Set Filter Function Lattice
+
+open Topology Filter Pointwise
+
+namespace Filter
+
+namespace IsGroupBasis
+
+/-- A simpler constructor for `Filter.IsGroupBasis` for bases valued in
+`Subgroup` (or any `S` satisfying `SubgroupClass S G`). -/
+@[to_additive "A simpler constructor for `Filter.IsAddGroupBasis` for bases valued in
+`AddSubgroup` (or any `S` satisfying `AddSubgroupClass S G`)."]
+theorem mk_of_subgroups {G S : Type*} {ι : Sort*} [Group G] [SetLike S G] [SubgroupClass S G]
+    {p : ι → Prop} {B : ι → S} (isBasis : IsBasis p (fun i ↦ B i : ι → Set G))
+    (conj : ∀ x₀, ∀ {i}, p i → ∃ j, p j ∧ MapsTo (x₀ * · * x₀⁻¹) (B j) (B i : Set G)) :
+    IsGroupBasis p (fun i ↦ B i : ι → Set G) where
+  toIsBasis := isBasis
+  one _ := one_mem _
+  mul {i} hi := ⟨i, hi, mul_subset_iff.mpr fun _ ha _ hb ↦ mul_mem ha hb⟩
+  inv {i} hi := ⟨i, hi, fun _ ha ↦ inv_mem ha⟩
+  conj := conj
+
+/-- A version of `Filter.IsGroupBasis.mk_of_subgroups` for commutative groups. -/
+@[to_additive "A version of `Filter.IsAddGroupBasis.mk_of_subgroups` for commutative groups."]
+theorem mk_of_subgroups_of_comm {G S : Type*} {ι : Sort*} [CommGroup G]
+    [SetLike S G] [SubgroupClass S G] {p : ι → Prop} {B : ι → S}
+    (isBasis : IsBasis p (fun i ↦ B i : ι → Set G)) :
+    IsGroupBasis p (fun i ↦ B i : ι → Set G) :=
+  .mk_of_comm _ _ isBasis (fun _ ↦ one_mem _)
+    (fun {i} hi ↦ ⟨i, hi, mul_subset_iff.mpr fun _ ha _ hb ↦ mul_mem ha hb⟩)
+    (fun {i} hi ↦ ⟨i, hi, fun _ ha ↦ inv_mem ha⟩)
+
+variable {G S : Type*} {ι : Sort*} [Group G] [SetLike S G] [SubgroupClass S G]
+  {p : ι → Prop} {B : ι → S} (hB : IsGroupBasis p (fun i ↦ B i : ι → Set G))
+include hB
+
+-- TODO(Anatole) : this should be a general lemma assuming `(𝓝 1).HasBasis p s`
```

Comment:
What is the obstruction to such a lemma?

Reply:
Mostly that I was lazy to think of the proper API and find the correct file to put this in, so I just stuck to something like the current API...